### PR TITLE
Add windows build to ci

### PR DESCRIPTION
**Pull request by Levi-Armstrong**

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
@johnwason I took your implementation for integrating windows ci and made a few modifications. Can you take a look?

File: .travis.yml [hunk 1]
```diff
@@ -1,25 +1,55 @@
...
+    - os: windows
+      language: cpp
 
 install:
-  - git clone --quiet --depth=1 https://github.com/ros-industrial/industrial_ci.git .ci_config
+  - if [ "$TRAVIS_OS_NAME" != "windows" ]; then git clone --quiet --depth=1 https://github.com/ros-industrial/industrial_ci.git .ci_config; fi
+  - |
+    if [ "$TRAVIS_OS_NAME" = "windows" ]; then
+        unset CXX
+        CXX_FOR_BUILD
```
Comment (johnwason) [inline .travis.yml]: this should be `unset CXX_FOR_BUILD`

**Comment by johnwason** [comment]
Assuming the Travis CI build works it should be fine. I'd suggest moving the setup and batch files into the centralized CI test repo since it will be the same for all projects.

File: .travis.yml (hunk 1 above)
Comment (Levi-Armstrong) [reply .travis.yml]: fixed

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn Where do you think I should put the repository for holding the two `.bat` files for building on windows?

**Comment by Levi-Armstrong** [comment]
@johnwason I moved the two .bat file to my gist for now instead of creating a repository two host just these two file. Does any one see any issues with this?

**Comment by johnwason** [comment]
If you want to store it in a gist you could combine the two scripts into a single file and have a command line option to select "install" and "build". Batch file syntax is a little... different but it shouldn't be too difficult.

This is untested but should give the general idea

    IF %1 == "install" (
       rem do install
       exit /b 0
    )

   IF %1 == "build" (
       rem do install
       exit /b 0
    )

    echo "Unknown command"
    exit /b 1

**Comment by gavanderhoorn** [comment]
> @gavanderhoorn Where do you think I should put the repository for holding the two `.bat` files for building on windows?

We could create a repository on `ros-industrial`.

It might also be nice to see whether we could upstream this to fi `industrial_ci`.

@ipa-mdl: have you considered adding Windows support?

File: .travis.yml [hunk 2]
```diff
@@ -1,25 +1,56 @@
...
 
 install:
-  - git clone --quiet --depth=1 https://github.com/ros-industrial/industrial_ci.git .ci_config
+  - if [ "$TRAVIS_OS_NAME" != "windows" ]; then git clone --quiet --depth=1 https://github.com/ros-industrial/industrial_ci.git .ci_config; fi
+  - |
+    if [ "$TRAVIS_OS_NAME" = "windows" ]; then
+        unset CXX
+        unset CXX_FOR_BUILD
+        unset CC
+        unset CC_FOR_BUILD
```
Comment (gavanderhoorn) [inline .travis.yml]: `unset` supports multiple variable names as input. These separate lines should not be necessary.

```batch
unset CXX CXX_FOR_BUILD CC CC_FOR_BUILD
```

should also work.

File: .travis.yml [hunk 3]
```diff
@@ -1,25 +1,56 @@
...
+        choco feature disable -n=showDownloadProgress
+        cd ..
+        mkdir -p catkin_ws/src
+        mv opw_kinematics catkin_ws/src/
+        cd catkin_ws
+        choco uninstall mingw
+        choco source add -n=ros-win -s="https://roswin.azurewebsites.net/api/v2" --priority=1
+        choco install ros-melodic-ros_base -y --execution-timeout=0
```
Comment (gavanderhoorn) [inline .travis.yml]: If we use `rosdep` later (in `travis_ci_windows.bat`), do we need to install `ros-melodic-ros_base` here?

File: include/opw_kinematics/opw_parameters_examples.h [hunk 4]
```diff
@@ -24,8 +33,7 @@ Parameters<T> makeIrb2400_10()
 }
 
 template <typename T>
-__attribute__((deprecated("UN-TESTED")))
-Parameters<T> makeFanucR2000iB_200R()
+DEPRECATED Parameters<T> makeFanucR2000iB_200R()
```
Comment (gavanderhoorn) [inline include/opw_kinematics/opw_parameters_examples.h]: It seems `DEPRECATED` does not support specifying a message.

Should not be hard to add (just do something like `DEPRECATED(x)` and pass `x` on if the target platform supports it.

File: .travis.yml (hunk 2 above)
Comment (Levi-Armstrong) [reply .travis.yml]: I will make the change.

File: .travis.yml (hunk 3 above)
Comment (Levi-Armstrong) [reply .travis.yml]: I will test it.

Comment (johnwason) [reply .travis.yml]: I doubt installing `ros-melodic-ros_base` can be skipped. It installs a mini Linux runtime environment with a huge number of dependencies that outside of "ROS" itself.

File: include/opw_kinematics/opw_parameters_examples.h (hunk 4 above)
Comment (Levi-Armstrong) [reply include/opw_kinematics/opw_parameters_examples.h]: I updated deprecated to take a message.

**Comment by Levi-Armstrong** [comment]
@johnwason I tried the [combined file](https://gist.github.com/Levi-Armstrong/e3bb008e579a99c52eb74235dff6735a#file-travis_ci_windows_combined-bat) approach but CI now keeps terminating because no output is recieved. Do you see any issues?

Also I had to remove the `"` to stop getting a syntax error but then ran into the issue previously described:
`cmd "/C call travis_ci_windows_install.bat"`
to
`cmd /C call travis_ci_windows_combined.bat build`

File: .travis.yml (hunk 2 above)
Comment (Levi-Armstrong) [reply .travis.yml]: It works.

File: .travis.yml (hunk 3 above)
Comment (Levi-Armstrong) [reply .travis.yml]: I tested it and it does not work because it could not find rosdep.

**Comment by johnwason** [comment]
It looks like if you add a space after the argument it works. Try this:

    cmd "/C call travis_ci_windows_combined.bat build "

**Comment by Levi-Armstrong** [comment]
@johnwason The extra space solved the problem.

File: .travis.yml (hunk 3 above)
Comment (gavanderhoorn) [reply .travis.yml]: You should be able to install  `rosdep` separately.

Comment (johnwason) [reply .travis.yml]: The `ros_base` package installs its own version of Python separate from the system version along with a lot of other support libraries and utilities. Untangling that dependency chain is probably not worth the effort.

**Comment by Levi-Armstrong** [comment]
It looks like everything is working now. It currently pulls a single gist file to build and install. @gavanderhoorn and @johnwason is there anything else that needs to be addressed before merging?

**Comment by johnwason** [comment]
Looks ok to me

**Comment by Levi-Armstrong** [comment]
@johnwason Is there anything that needs to be done for the shared libraries on windows build?

**Comment by johnwason** [comment]
This repo appears to be header only without libraries being built.

**Comment by johnwason** [comment]
See #34 for a cleaner option for Windows builds

**Pull request merged**